File: ProdCons3.java
// Liam Wynn, 1/26/2021, CS515p: Assignment 1

/*
  This is the Java version of prodcons.cpp.
*/

import java.util.LinkedList;
import java.util.Queue;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.locks.Condition;

class ProdCons3 {
  public static final int BUFSIZE = 20;
  public static final int NUMITEMS = 100;
  public static final int TERMINATE = -1;

  public static final Lock queueLock = new ReentrantLock();
  public static final Condition queueNotFull  = queueLock.newCondition();
  public static final Condition queueNotEmpty = queueLock.newCondition();

  public static final Object consumerStatsLck = new Object();

  public static int numProducers;
  public static int itemsPerProducer;
  public static int itemsRemainder;

  // A value each producer will increment when they
  // finish. Used to identify the last producer, who
  // will add the TERMINATE values to the queue.
  public static int producerFinishedCount;
  public static final Object prodFinLck = new Object();

  public static void main(String[] args) {
    int numConsumers = 1;

    Producer[] producerThreads;
    Consumer[] consumerThreads;

    Queue<Integer> queue = new LinkedList<>();
    int[] consumerStats;

    try {
      if(args.length >= 2) {
        numConsumers = Integer.parseInt(args[0]);
        numProducers = Integer.parseInt(args[1]);
      } else if(args.length == 1)
        numConsumers = Integer.parseInt(args[0]);
    } catch(Exception e) {
      numConsumers = 1;
      numProducers = 1;
    }

    numConsumers = numConsumers <= 0 ? 1 : numConsumers;
    numProducers = numProducers <= 0 ? 1 : numProducers;

    itemsPerProducer = NUMITEMS / numProducers;
    itemsRemainder = NUMITEMS - (itemsPerProducer * numProducers);

    producerThreads = new Producer[numProducers];
    for(int i = 0; i < numProducers; i++) {
      if(i == numProducers - 1)
        producerThreads[i] = new Producer(i, numConsumers, queue);
      else
        producerThreads[i] = new Producer(i, numConsumers, queue);
    }

    consumerThreads = new Consumer[numConsumers];
    consumerStats = new int[numConsumers];
    for(int i = 0; i < numConsumers; i++) {
      consumerThreads[i] = new Consumer(i, queue, consumerStats);
    }

    try {
      for(int i = 0; i < numProducers; i++)
        producerThreads[i].start();
      for(int i = 0; i < numConsumers; i++)
        consumerThreads[i].start();

      for(int i = 0; i < numProducers; i++)
        producerThreads[i].join();
      for(int i = 0; i < numConsumers; i++) {
        consumerThreads[i].join();
      }

      int total = 0;
      System.out.print("consumer stats: [");
      for(int i = 0; i < numConsumers; i++) {
         total += consumerStats[i];
         System.out.print(consumerStats[i] + ",");
      }
      System.out.println("], total = " + total);
  
    } catch(Exception e) {
      System.out.println("Error running threads");
    }
  }
}

class Producer extends java.lang.Thread {
  int id;
  // Need this to know how many TERMINATES to add to the queue.
  private int numConsumers;
  private Queue<Integer> queue;

  public Producer(int id, int numConsumers, Queue<Integer> queue) {
    this.id = id;
    this.numConsumers = numConsumers;
    this.queue = queue;
  }

  public void run() {
    int start = id * ProdCons3.itemsPerProducer + 1;
    int end = start + ProdCons3.itemsPerProducer - 1;

    if(id == ProdCons3.numProducers - 1)
      end += ProdCons3.itemsRemainder;

    System.out.println("producer[" + id + "] for segment [" + start + ", " + end + "] is running");

    for(int i = start; i <= end; i++) {
      ProdCons3.queueLock.lock();

      try {
		while(queue.size() == ProdCons3.BUFSIZE)
          ProdCons3.queueNotFull.await();

        queue.add(i);
        System.out.println("producer[" + id + "] added " + i + "(qsz: " + queue.size() + ")");

        ProdCons3.queueNotEmpty.signal();

      } catch(Exception e) {
        System.out.println(e.getMessage());
      } finally {
        ProdCons3.queueLock.unlock();
      }
    }

    System.out.println("producer[" + id + "] ending");
    synchronized(ProdCons3.prodFinLck) {
      ProdCons3.producerFinishedCount++;

      if(ProdCons3.producerFinishedCount == ProdCons3.numProducers) {
        for(int i = 0; i <= numConsumers; i++) {
          ProdCons3.queueLock.lock();

          try {
	    	while(queue.size() == ProdCons3.BUFSIZE)
              ProdCons3.queueNotFull.await();

            queue.add(-1);
            ProdCons3.queueNotEmpty.signal();

          } catch(Exception e) {
            System.out.println(e.getMessage());
          } finally {
            ProdCons3.queueLock.unlock();
          }
        }
      }
    }
  }
}

class Consumer extends java.lang.Thread {
  private int id;
  private int count;

  private Queue<Integer> queue;
  private int[] stats;

  public Consumer(int id, Queue<Integer> queue, int[] stats) {
    this.id = id;
    this.count = 0;

    this.queue = queue;
    this.stats = stats;
  }

  public void run() {
    for(int i = 1; i < ProdCons3.NUMITEMS; i++) {
      ProdCons3.queueLock.lock();

      try {
        while(queue.size() == 0)
          ProdCons3.queueNotEmpty.await();

        int p = queue.remove();
        if(p == ProdCons3.TERMINATE) {
          ProdCons3.queueNotFull.signal();
          ProdCons3.queueLock.unlock();
          break;
        } else {
          count++;
          System.out.println("consumer[" + id + "] rem'd " + p + " (qsz: " + queue.size() + ")");
          ProdCons3.queueNotFull.signal();
        }
      } catch(Exception e) {
        ProdCons3.queueLock.unlock();
        System.out.println(e.getMessage());
      }

      ProdCons3.queueLock.unlock();
    }

    System.out.println("consumer[" + id + "] ending");

    synchronized(ProdCons3.consumerStatsLck) {
      stats[id] = count;
    }
  }
}

File: prodcons3.cpp
// Liam Wynn, 1/19/2021, CS515p: Parallel Programming

/*
  In this assignment we implement a full
  rich producer-consumer model. Whereas
  the previous iteration had a variable
  number of consumers, this one will have 
  also a variable number of producers.
  Each producer will be responsible for
  a portion of all the items.
*/

#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <vector>
#include "./queue.h"

using namespace std;

// The queue's capacity.
const int BUFSIZE = 20;
// We add all values numbers from 1
// to this value into the queue.
const int NUMITEMS = 100;
// When read, it tells a consumer to stop
// consuming.
const int TERMINATE_VAL = -1;

// The number of threads that will add items
// to the queue.
int num_producers;
// The number of items for each producer thread.
// This will be NUMITEMS / num_producers.
int items_per_thread;
// If NUMITEMS / num_producers has a remainder,
// we store that here. This will be added to the
// last producer thread.
int items_remainder;

// Make global since producer and
// main thread handle this.
int num_consumers;
// Keeps track of each consumer thread's items they
// consumed.
vector<int> consumed_count;

Queue queue(BUFSIZE);
mutex mtx_queue_state;
condition_variable queue_not_empty, queue_not_full;

// Mutex for controlling output. cout is not atomic!
mutex mtx_output;

// When a producer thread is done, it will update
// this value. When the last thread reaches this
// value, it will add the terminating values.
mutex mtx_producer_finished;
int prod_finished_count;

void producer(const int thread_id) {
  int start, end;

  // Plus 1 so that we start at 1.
  start = thread_id * items_per_thread + 1;
  end = start + items_per_thread - 1;

  if(thread_id == num_producers - 1)
    end += items_remainder;

  mtx_output.lock();
  cout << "Producer[" << thread_id << "] for segment ["
       << start << "," << end << "] starting on core " << sched_getcpu() << endl;
  mtx_output.unlock();

  // Each producer can had its items in any order.
  for(int i = start; i <= end; i++) {
    unique_lock<mutex> queue_lck(mtx_queue_state);
    while(queue.size() == BUFSIZE) {
      queue_not_full.wait(queue_lck);
    }

    // Once we finish adding normal values,
    // we add terminate value to tell consumers
    // to stop running.
    queue.add(i);

    mtx_output.lock();
    cout << "Producer[" << thread_id << "] added " << i << " (qsz: " << queue.size() << ")" << endl;
    mtx_output.unlock();

    queue_not_empty.notify_one();
    queue_lck.unlock();
  }

  mtx_output.lock();
  cout << "Producer[" << thread_id << "] ending" << endl;
  mtx_output.unlock();

  mtx_producer_finished.lock();
  prod_finished_count++;
  if(prod_finished_count == num_producers) {
    for(int i = 0; i < num_consumers; i++) { 
      unique_lock<mutex> queue_lck(mtx_queue_state);
      while(queue.size() == BUFSIZE)
        queue_not_full.wait(queue_lck);

      queue.add(TERMINATE_VAL);
      queue_not_empty.notify_one();
      queue_lck.unlock();
    }
  }
  mtx_producer_finished.unlock();
}

void consumer(const int thread_id) {
  mtx_output.lock();
  cout << "Consumer[" << thread_id << "] starting on core " << sched_getcpu() << endl;
  mtx_output.unlock();

  for(int i = 1; i <= NUMITEMS; i++) {
    unique_lock<mutex> queue_lck(mtx_queue_state);

    while(queue.isEmpty())
      queue_not_empty.wait(queue_lck);

    int item = queue.remove();
    if(item == TERMINATE_VAL) {
      queue_not_full.notify_one();
      queue_lck.unlock();
      break;
    }

    consumed_count[thread_id] += 1;

    mtx_output.lock();
    cout << "Consumer[" << thread_id << "] rem'd " << item << " (qsz: " << queue.size() << ")" << endl;
    mtx_output.unlock();

    queue_not_full.notify_one();
    queue_lck.unlock();
  }

  mtx_output.lock();
  cout << "Consumer[" << thread_id << "] ending" << endl;
  mtx_output.unlock();
}

int main(int argc, char** argv) {
  num_consumers = 1;
  num_producers = 1;
  prod_finished_count = 0;

  if(argc >= 3) {
    num_consumers = atoi(argv[1]);
    num_producers = atoi(argv[2]);
  } else if(argc >= 2) {
    num_consumers = atoi(argv[1]);
  }

  num_consumers = num_consumers <= 0 ? 1 : num_consumers;
  num_producers = num_producers <= 0 ? 1 : num_producers;

  items_per_thread = NUMITEMS / num_producers;
  items_remainder = NUMITEMS - (items_per_thread * num_producers);

  cout << "Number of consumers is " << num_consumers << endl;
  cout << "Number of producers is " << num_producers << endl;

  vector<thread> producers;
  vector<thread> consumers;

  for(int i = 0; i < num_producers; i++)
    producers.push_back(thread(producer, i));

  for(int i = 0; i < num_consumers; i++) {
    consumed_count.push_back(0);
    consumers.push_back(thread(consumer, i));
  }

  for(auto t = producers.begin(); t != producers.end(); t++)
    (*t).join();
  for(auto t = consumers.begin(); t != consumers.end(); t++)
    (*t).join();

  int total = 0;
  cout << "Consumer stats: ";
  cout << "[";
  for(auto t = consumed_count.begin(); t != consumed_count.end(); t++) {
    total += *t;
    cout << *t << ", ";
  }
  cout << "] = " << total << endl;

  cout << "main: all done!" << endl;
  return 0;
}
